In [2]:
# ===== 기본 세트 (거의 항상 필요) =====
import pandas as pd                    # 데이터프레임 다루기 (표 형태 데이터 처리)
import numpy as np                     # 수치 계산, 배열 연산
import matplotlib.pyplot as plt        # 기본 시각화
import matplotlib.font_manager as fm   # 그래프에 한글 폰트 적용 시 필요
import seaborn as sns                  # matplotlib 기반의, 더 예쁘고 간결한 통계 시각화

# ===== 한글 폰트 설정 =====
font_path = "C:/Windows/Fonts/malgun.ttf"
fm.fontManager.addfont(font_path)
plt.rcParams["font.family"] = fm.FontProperties(fname=font_path).get_name()
plt.rcParams["axes.unicode_minus"] = False

# ===== 파일/경로/시스템 =====
import os                              # 파일 경로 확인, 파일 존재 여부(os.path.exists) 등
from pathlib import Path               # 폴더/파일 경로를 객체로 다루기 (work_dir / "파일명")
import joblib                          # 학습된 scaler/encoder를 파일로 저장하고 재사용
import platform                        # OS 종류 확인 (Windows/Mac/Linux 구분 필요할 때)
import warnings                        # 경고 메시지 제어(숨기기/표시)
import matplotlib.pyplot as plt
import urllib.request, zipfile

# ===== 벤치마크/성능 측정 =====
import time                            # 코드 실행 시간 측정 (time.perf_counter())
import psutil                          # 메모리 사용량 측정 (rss_mb 등)
import gc                              # 메모리 강제 정리(garbage collection)

# ===== 결측치 탐색 시각화 =====
import missingno as msno               # 결측치 위치(matrix), 관계(heatmap) 시각화

# ===== 인터랙티브 시각화 =====
import plotly.express as px            # 마우스 hover/zoom 가능한 인터랙티브 그래프
import plotly.io as pio                # plotly 렌더러(출력 방식) 설정

# ===== 대용량 데이터 처리 =====
import polars as pl                    # pandas보다 빠른 대용량 데이터 처리 (Rust 기반)

# ===== 통계 검증 =====
import statsmodels.api as sm           # 회귀분석, 다중공선성(Cond. No.) 등 통계 검증
import statsmodels.formula.api as smf  # R 스타일 수식("y ~ x1 + x2")으로 회귀분석
from scipy import stats                # t-검정·카이제곱·정규성 검정 등 통계 검정 모음
from statsmodels.stats.power import TTestIndPower   # 검정력 분석(필요 표본수, 검정력 계산)
from sklearn.neighbors import NearestNeighbors      # 최근접 이웃 탐색 (경향점수 매칭 등에 사용)

# ===== 전처리(인코딩/스케일링) =====
from sklearn.preprocessing import LabelEncoder      # 카테고리를 숫자로 변환 (순서형)
from sklearn.preprocessing import MinMaxScaler      # 0~1 범위로 스케일링
from sklearn.preprocessing import StandardScaler    # 평균0, 표준편차1로 스케일링
from sklearn.preprocessing import RobustScaler      # 이상치에 강한 스케일링 (median/IQR 기반)

# ===== 모델 =====
from sklearn.linear_model import LinearRegression       # 선형 회귀 모델
from sklearn.linear_model import LogisticRegression     # 로지스틱 회귀(분류) 모델
from sklearn.tree import DecisionTreeClassifier         # 의사결정나무 분류 모델
from sklearn.dummy import DummyClassifier               # 베이스라인(단순 규칙) 분류 모델
from sklearn.ensemble import RandomForestRegressor      # 랜덤포레스트 회귀 모델
from sklearn.ensemble import RandomForestClassifier     # 랜덤포레스트 분류 모델(여러 트리의 앙상블)
from xgboost import XGBClassifier                       # XGBoost 분류 모델(부스팅 기반)

# ===== 파이프라인 =====
from sklearn.pipeline import make_pipeline              # 전처리+모델 여러 단계를 하나로 묶기

# ===== 데이터 분리 / 교차검증 =====
from sklearn.model_selection import train_test_split    # 데이터를 학습용/테스트용으로 분리
from sklearn.model_selection import KFold               # 데이터를 K개로 분할(교차검증용, 회귀에 적합)
from sklearn.model_selection import StratifiedKFold     # 클래스 비율을 유지하며 K개로 분할(불균형 분류용)
from sklearn.model_selection import cross_val_score     # 교차검증으로 단일 지표 반복 평가
from sklearn.model_selection import cross_validate      # 교차검증으로 여러 지표·학습점수까지 한번에 평가
from sklearn.model_selection import learning_curve      # 표본 크기에 따른 학습/검증 성능 변화 확인

# ===== 평가지표 =====
from sklearn.metrics import accuracy_score              # 분류 모델의 정확도 계산
from sklearn.metrics import confusion_matrix            # 실제/예측 결과를 교차표로 정리(TP/FP/TN/FN)
from sklearn.metrics import classification_report       # 정밀도·재현율·F1 등을 한번에 요약
from sklearn.metrics import precision_score             # 정밀도(예측한 것 중 실제로 맞은 비율)
from sklearn.metrics import recall_score                # 재현율(실제 정답 중 모델이 잡아낸 비율)
from sklearn.metrics import f1_score                     # 정밀도와 재현율의 조화평균
from sklearn.metrics import roc_curve                    # ROC 곡선 좌표 계산
from sklearn.metrics import roc_auc_score                # ROC 곡선 아래 면적(AUC)
from sklearn.metrics import precision_recall_curve       # 정밀도-재현율 곡선 좌표 계산
from sklearn.metrics import average_precision_score      # PR 곡선 아래 면적(AP)
from sklearn.metrics import log_loss                      # 분류 확률 예측의 손실(로그 손실)

In [3]:
DATA_URL = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/"

RANDOM_STATE = 42
np.random.seed(RANDOM_STATE)

# 한글 폰트 설정 — 그래프의 한글이 □□□로 깨지지 않게 합니다
system = platform.system()
if system == "Darwin":                      # macOS
    plt.rcParams["font.family"] = "AppleGothic"
elif system == "Windows":
    plt.rcParams["font.family"] = "Malgun Gothic"
else:                                        # 코랩 등 리눅스 — 나눔고딕을 설치해야 합니다
    if not any("Nanum" in f.name for f in fm.fontManager.ttflist):
        print("나눔고딕을 설치합니다 (10~20초 걸립니다)...")
        import subprocess
        subprocess.run("apt-get install -y -qq fonts-nanum", shell=True,
                       capture_output=True)
        for path in fm.findSystemFonts(fontpaths=["/usr/share/fonts/truetype/nanum"]):
            fm.fontManager.addfont(path)
    plt.rcParams["font.family"] = "NanumGothic"

plt.rcParams["axes.unicode_minus"] = False
sns.set_theme(style="whitegrid", font=plt.rcParams["font.family"])

print("환경 준비 완료!")
import sklearn
print(f"- scikit-learn {sklearn.__version__}")
print(f"- 그래프 한글 폰트: {plt.rcParams['font.family'][0]}")

환경 준비 완료!
- scikit-learn 1.9.0
- 그래프 한글 폰트: Malgun Gothic


In [4]:
titanic = pd.read_csv(DATA_URL + "titanic.csv")

# 오늘 쓸 컬럼만 골라냅니다
df = titanic[["survived", "pclass", "sex", "age", "sibsp", "parch",
              "fare", "embarked", "who"]].copy()

print(f"전체 {df.shape[0]}명 × 컬럼 {df.shape[1]}개")
print(f"생존율 {df['survived'].mean():.4f}   (타깃: survived)")
print()
print(df.head(5).to_string())
print()
print("타입별 컬럼")
print(f"  수치형: {df.select_dtypes(include='number').columns.tolist()}")
print(f"  범주형: {df.select_dtypes(exclude='number').columns.tolist()}")
print()
na = df.isna().sum()
print("결측 있는 컬럼")
print(na[na > 0].to_string())
print()
print("→ age에 결측이 177건 있습니다. 이것을 어떻게 채울지도 오늘의 주제입니다.")

전체 891명 × 컬럼 9개
생존율 0.3838   (타깃: survived)

   survived  pclass     sex   age  sibsp  parch     fare embarked    who
0         0       3    male  22.0      1      0   7.2500        S    man
1         1       1  female  38.0      1      0  71.2833        C  woman
2         1       3  female  26.0      0      0   7.9250        S  woman
3         1       1  female  35.0      1      0  53.1000        S  woman
4         0       3    male  35.0      0      0   8.0500        S    man

타입별 컬럼
  수치형: ['survived', 'pclass', 'age', 'sibsp', 'parch', 'fare']
  범주형: ['sex', 'embarked', 'who']

결측 있는 컬럼
age         177
embarked      2

→ age에 결측이 177건 있습니다. 이것을 어떻게 채울지도 오늘의 주제입니다.


In [5]:
# ─────────────────────────────────────────────
# [C3] Part 1 · 📊 데이터로 확인해 봅시다 — 만들고, 바꾸고, 맞춰 보기
# 파생변수 3개 생성 → 가설 검증 (실행하면 가족 크기별 생존율이 나옵니다)
# ─────────────────────────────────────────────
# pandas로 만들던 그 조작입니다
df["family_size"] = df["sibsp"] + df["parch"] + 1
df["is_alone"] = (df["family_size"] == 1).astype(int)
df["fare_per_person"] = (df["fare"] / df["family_size"]).round(2)

print("가족 크기별 생존율")
fam = df.groupby("family_size")["survived"].agg(["mean", "size"])
fam.columns = ["생존율", "인원"]
print(fam.round(3).to_string())
print()
print(f"혼자 탄 사람({int(df['is_alone'].sum())}명) 생존율 : "
      f"{df.loc[df['is_alone'] == 1, 'survived'].mean():.3f}")
print(f"가족과 탄 사람({int((1 - df['is_alone']).sum())}명) 생존율: "
      f"{df.loc[df['is_alone'] == 0, 'survived'].mean():.3f}")

가족 크기별 생존율
               생존율   인원
family_size            
1            0.304  537
2            0.553  161
3            0.578  102
4            0.724   29
5            0.200   15
6            0.136   22
7            0.333   12
8            0.000    6
11           0.000    7

혼자 탄 사람(537명) 생존율 : 0.304
가족과 탄 사람(354명) 생존율: 0.506


In [6]:
# ─────────────────────────────────────────────
# [C4] Part 1 · 📊 데이터로 확인해 봅시다 — 만들고, 바꾸고, 맞춰 보기
# One-Hot vs Ordinal — embarked를 두 방식으로 인코딩해 비교 (실행하면 CV 점수가 나옵니다)
# ─────────────────────────────────────────────
from sklearn.preprocessing import OneHotEncoder
from sklearn.preprocessing import OrdinalEncoder
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline
from sklearn.impute import SimpleImputer

y = df["survived"]
cv = StratifiedKFold(n_splits=5, shuffle=True, random_state=RANDOM_STATE)
base_num = ["pclass", "age", "fare"]      # 효과를 보려고 일부러 최소 구성으로

encoders = [
    ("One-Hot", OneHotEncoder(handle_unknown="ignore")),
    ("Ordinal(Label)", OrdinalEncoder(handle_unknown="use_encoded_value", unknown_value=-1)),
]

for label, enc in encoders:
    pre = ColumnTransformer([
        ("num", Pipeline([("impute", SimpleImputer(strategy="median")),
                          ("scale", StandardScaler())]), base_num),
        ("cat", Pipeline([("impute", SimpleImputer(strategy="most_frequent")),
                          ("encode", enc)]), ["embarked"]),
    ])
    pipe = Pipeline([("pre", pre),
                     ("clf", LogisticRegression(max_iter=5000, random_state=RANDOM_STATE))])
    s = cross_val_score(pipe, df[base_num + ["embarked"]], y, cv=cv, scoring="accuracy")
    print(f"  {label:16} CV 정확도 {s.mean():.4f} ± {s.std():.4f}")

print()
print(f"embarked의 범주: {sorted(df['embarked'].dropna().unique().tolist())} ({df['embarked'].nunique()}개)")

  One-Hot          CV 정확도 0.7003 ± 0.0231
  Ordinal(Label)   CV 정확도 0.7104 ± 0.0333

embarked의 범주: ['C', 'Q', 'S'] (3개)


In [7]:
# ─────────────────────────────────────────────
# [C5] Part 1 · 📊 데이터로 확인해 봅시다 — 만들고, 바꾸고, 맞춰 보기
# 스케일링이 거리 기반 모델(KNN)에 미치는 영향 (실행하면 스케일러별 점수가 나옵니다)
# ─────────────────────────────────────────────
from sklearn.neighbors import KNeighborsClassifier
NUM_COLS = ["pclass", "age", "sibsp", "parch", "fare",
            "family_size", "is_alone", "fare_per_person"]

print("컬럼별 값의 범위 (스케일 차이 확인)")
rng = df[NUM_COLS].agg(["min", "max"]).T
rng["범위"] = rng["max"] - rng["min"]
print(rng.round(1).to_string())
print()

print("KNN(n_neighbors=5) 성능 — 스케일러별")
scalers = [(None, "스케일링 없음"), (StandardScaler(), "StandardScaler"),
           (MinMaxScaler(), "MinMaxScaler"), (RobustScaler(), "RobustScaler")]
for sc, label in scalers:
    steps = [("impute", SimpleImputer(strategy="median"))]
    if sc is not None:
        steps.append(("scale", sc))
    steps.append(("clf", KNeighborsClassifier(n_neighbors=5)))
    s = cross_val_score(Pipeline(steps), df[NUM_COLS], y, cv=cv, scoring="accuracy")
    print(f"  {label:16} {s.mean():.4f} ± {s.std():.4f}")

컬럼별 값의 범위 (스케일 차이 확인)
                 min    max     범위
pclass           1.0    3.0    2.0
age              0.4   80.0   79.6
sibsp            0.0    8.0    8.0
parch            0.0    6.0    6.0
fare             0.0  512.3  512.3
family_size      1.0   11.0   10.0
is_alone         0.0    1.0    1.0
fare_per_person  0.0  512.3  512.3

KNN(n_neighbors=5) 성능 — 스케일러별
  스케일링 없음          0.6958 ± 0.0180
  StandardScaler   0.7059 ± 0.0335
  MinMaxScaler     0.7014 ± 0.0336
  RobustScaler     0.7104 ± 0.0226


In [8]:
# [C6] Part 1 · ⌨️ 백문이 불여일타! (1)
# ⌨️ 백문이 불여일타! (1) — age를 구간화해 새 피처 만들고 가설 검증

bins = [0, 12, 19, 59, 200]
labels = ["child(0~12)", "teen(13~19)", "adult(20~59)", "senior(60~)"]

# 여기에 코드를 작성하세요

#[문제]
#파생변수를 하나 더 만들어 가설을 검증합니다.

#1) age를 구간으로 나눈 age_group을 만듭니다.
#   구간: 0~12(child) / 13~19(teen) / 20~59(adult) / 60~(senior)
#   힌트: pd.cut()을 쓰면 한 줄입니다. 결측은 그대로 둡니다.
#2) age_group별 생존율과 인원을 출력합니다.
#3) 이 파생변수가 '좋은 피처의 조건' 세 가지를 만족하는지 판단합니다.

# 1) age_group 만들기
df["age_group"] = pd.cut(df["age"], bins=bins, labels=labels)

# 2) age_group별 생존율과 인원 확인
survival_by_group = df.groupby("age_group")["survived"].agg(["mean", "count"])
survival_by_group.columns = ["생존율", "인원수"]
print(survival_by_group.round(3))
print("\n 나이가 많아질수록 생존율이 하락하는 패턴은 보이나 adult의 인원수가 너무 많고 senior의 표본의 수가 너무 적어 좋은 피쳐로 쓰기는 어려움")


                생존율  인원수
age_group               
child(0~12)   0.580   69
teen(13~19)   0.411   95
adult(20~59)  0.389  524
senior(60~)   0.269   26

 나이가 많아질수록 생존율이 하락하는 패턴은 보이나 adult의 인원수가 너무 많고 senior의 표본의 수가 너무 적어 좋은 피쳐로 쓰기는 어려움


In [9]:
# ─────────────────────────────────────────────
# [C7] Part 2 · 📊 데이터로 확인해 봅시다 — 검증 누수를 재현하기
# (나쁜 방법) 테스트로 고르기 vs (옳은 방법) CV로 고르기 (실행하면 두 결론이 비교됩니다)
# ─────────────────────────────────────────────
FEAT_NUM = NUM_COLS                       # [C5]에서 정한 수치형 8개
FEAT_CAT = ["sex", "embarked", "who"]
X = df[FEAT_NUM + FEAT_CAT]

X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.25, random_state=RANDOM_STATE, stratify=y)


def make_pre():
    """수치·범주 분기 전처리를 새로 만들어 돌려줍니다."""
    return ColumnTransformer([
        ("num", Pipeline([("impute", SimpleImputer(strategy="median")),
                          ("scale", StandardScaler())]), FEAT_NUM),
        ("cat", Pipeline([("impute", SimpleImputer(strategy="most_frequent")),
                          ("encode", OneHotEncoder(handle_unknown="ignore"))]), FEAT_CAT),
    ])


rows = []
for depth in [2, 3, 4, 5, 7, 10, None]:
    pipe = Pipeline([("pre", make_pre()),
                     ("clf", DecisionTreeClassifier(max_depth=depth,
                                                    random_state=RANDOM_STATE))])
    cv_score = cross_val_score(pipe, X_train, y_train, cv=cv, scoring="accuracy").mean()
    pipe.fit(X_train, y_train)
    test_score = pipe.score(X_test, y_test)
    rows.append({"max_depth": "제한 없음" if depth is None else depth,
                 "학습셋 CV": cv_score, "테스트": test_score})

sel = pd.DataFrame(rows)
print(sel.round(4).to_string(index=False))
print()
bad = sel.loc[sel["테스트"].idxmax()]
good = sel.loc[sel["학습셋 CV"].idxmax()]
print(f"✗ 나쁜 방법 — 테스트로 고름: max_depth={bad['max_depth']} → 보고 점수 {bad['테스트']:.4f}")
print(f"✓ 옳은 방법 — CV로 고름:    max_depth={good['max_depth']} → 테스트 점수 {good['테스트']:.4f}")
print(f"→ 부풀림 {bad['테스트'] - good['테스트']:+.4f}")

max_depth  학습셋 CV    테스트
        2  0.7949 0.7937
        3  0.8113 0.8072
        4  0.8053 0.8161
        5  0.8158 0.7623
        7  0.8069 0.7578
       10  0.7814 0.7489
    제한 없음  0.7620 0.7265

✗ 나쁜 방법 — 테스트로 고름: max_depth=4 → 보고 점수 0.8161
✓ 옳은 방법 — CV로 고름:    max_depth=5 → 테스트 점수 0.7623
→ 부풀림 +0.0538


In [11]:
# [C8] Part 2 · ⌨️ 백문이 불여일타! (2)
# ⌨️ 백문이 불여일타! (2) — 동료의 코드에서 누수 찾기

# (답을 코드로 검증해보고 싶다면 아래에 직접 재현해보세요)

# 여기에 코드를 작성하세요

  # 1) 전체 데이터로 결측을 채운다
X_filled = X.copy()
X_filled["age"] = X_filled["age"].fillna(X_filled["age"].median())
# 결측치 처리: train의 중앙값만 계산해서, train/test 둘 다에 적용

  # 2) 전체 데이터로 스케일링한다
X_scaled = StandardScaler().fit_transform(X_filled[FEAT_NUM])
# train으로만 fit
# test는 transform만

  # 3) 나눠서 학습한다
Xa, Xb, ya, yb = train_test_split(X_scaled, y, test_size=0.25, random_state=42)
model = DecisionTreeClassifier(max_depth=4).fit(Xa, ya)
print("테스트 정확도:", model.score(Xb, yb))

테스트 정확도: 0.7040358744394619


1. 데이터 누수의 네 가지 유형 (각 한 줄)
① 전처리 누수: 스케일링·결측치 대체 등의 "기준값(평균, 중앙값 등)"을 전체 데이터(train+test)로 계산해서, test의 정보가 train 쪽에 스며드는 경우. (오늘 직접 확인한 사례)
② 타겟 누수: 예측 시점에는 절대 알 수 없는, "미래 정보" 또는 "정답과 사실상 같은 정보"를 피처로 사용하는 경우 (예: "퇴원일"로 "입원 기간"을 예측하는 것).
③ 검증 누수: 피처 선택이나 하이퍼파라미터 튜닝을, 교차검증 fold로 나누기 "전에" 전체 데이터로 먼저 해버려서, 검증용 fold의 정보가 이미 모델 설계에 반영되는 경우.
④ 시간 누수: 시계열 데이터에서, 무작위로 섞어서 나누는 바람에 "미래 데이터"가 학습에 쓰이고 "과거 데이터"로 그걸 테스트하게 되는 경우 (현실에서는 절대 있을 수 없는 순서).

2. 누수를 의심해야 하는 신호
① 성능이 비정상적으로 좋을 때: 도메인 상식으로 봤을 때 "이 정도로 어려운 문제인데 정확도 99%가 나온다"면 의심해야 합니다.
② 학습 성능과 테스트 성능이 거의 차이가 없거나, 테스트가 오히려 더 좋을 때: 정상적으로는 테스트가 학습보다 약간 낮은 게 자연스러운데, 그 반대라면 뭔가 정보가 새고 있다는 신호입니다.
③ 특정 피처 하나의 중요도가 압도적으로 높을 때: "왜 이 피처가 이렇게까지 결정적이지?"라는 의문이 들면, 그 피처가 사실 타겟과 거의 동일한 정보를 담고 있는 건 아닌지 확인해야 합니다.
④ 실제 배포 후 성능이 테스트 때와 완전히 다르게 뚝 떨어질 때: 이건 "테스트 점수 자체가 이미 오염되어 있었다"는 사후적 증거입니다.

3. 왜 Pipeline으로도 "검증 누수(유형③)"는 못 막을까요
Pipeline은 "그 파이프라인 안에 명시적으로 포함된 단계들"만, 매 fold마다 올바른 순서(train으로 fit, test로 transform)로 자동 처리해줍니다. 하지만 "파이프라인을 만들기도 전에, 이미 전체 데이터로 끝내버린 작업"은, 파이프라인이 손댈 수 있는 범위 밖에 있어서 전혀 막아주지 못합니다.

In [12]:
# ─────────────────────────────────────────────
# [C9] Part 3 · 📊 데이터로 확인해 봅시다 — 묶고, 새는지 재보기
# 기본 Pipeline 만들기 → 내부 단계 확인 (실행하면 단계 목록과 학습된 기준값이 나옵니다)
# ─────────────────────────────────────────────
simple = Pipeline([
    ("impute", SimpleImputer(strategy="median")),
    ("scale", StandardScaler()),
    ("clf", LogisticRegression(max_iter=5000, random_state=RANDOM_STATE)),
])

simple.fit(X_train[FEAT_NUM], y_train)

print("Pipeline 단계 목록")
for name, step in simple.named_steps.items():
    print(f"  {name:8} {type(step).__name__}")
print()
print("named_steps로 내부에 접근할 수 있습니다")
med = simple.named_steps["impute"].statistics_
print(f"  대치에 쓴 중앙값 (앞 3개): {med[:3].round(2)}")
print(f"  → age 중앙값 {med[FEAT_NUM.index('age')]:.1f}  (학습 조각에서만 계산된 값)")
print()
print(f"테스트 정확도: {simple.score(X_test[FEAT_NUM], y_test):.4f}")
print()
print("→ predict할 때 전처리는 fit을 다시 하지 않고, 학습 때의 기준값을 '적용만' 합니다.")

Pipeline 단계 목록
  impute   SimpleImputer
  scale    StandardScaler
  clf      LogisticRegression

named_steps로 내부에 접근할 수 있습니다
  대치에 쓴 중앙값 (앞 3개): [ 3. 29.  0.]
  → age 중앙값 29.0  (학습 조각에서만 계산된 값)

테스트 정확도: 0.6816

→ predict할 때 전처리는 fit을 다시 하지 않고, 학습 때의 기준값을 '적용만' 합니다.


In [13]:
# ─────────────────────────────────────────────
# [C10] Part 3 · 📊 데이터로 확인해 봅시다 — 묶고, 새는지 재보기
# ColumnTransformer로 수치·범주 분기 → 두 겹 Pipeline 완성 (실행하면 변환 후 피처명이 나옵니다)
# ─────────────────────────────────────────────
model = Pipeline([("pre", make_pre()),
                  ("clf", LogisticRegression(max_iter=5000, random_state=RANDOM_STATE))])
model.fit(X_train, y_train)

names = model.named_steps["pre"].get_feature_names_out()
print(f"입력 컬럼 {X_train.shape[1]}개 → 변환 후 피처 {len(names)}개")
print()
print("변환 후 피처 이름")
for n in names:
    print(f"  {n}")
print()
print(f"테스트 정확도: {model.score(X_test, y_test):.4f}")
print()
print("→ 'num__'·'cat__' 접두사로 어느 갈래에서 왔는지 알 수 있습니다.")
print("  범주형 3개가 One-Hot으로 펼쳐져 열이 늘어났습니다.")

입력 컬럼 11개 → 변환 후 피처 16개

변환 후 피처 이름
  num__pclass
  num__age
  num__sibsp
  num__parch
  num__fare
  num__family_size
  num__is_alone
  num__fare_per_person
  cat__sex_female
  cat__sex_male
  cat__embarked_C
  cat__embarked_Q
  cat__embarked_S
  cat__who_child
  cat__who_man
  cat__who_woman

테스트 정확도: 0.8161

→ 'num__'·'cat__' 접두사로 어느 갈래에서 왔는지 알 수 있습니다.
  범주형 3개가 One-Hot으로 펼쳐져 열이 늘어났습니다.


In [23]:
# ─────────────────────────────────────────────
# [C11] Part 3 · 📊 데이터로 확인해 봅시다 — 묶고, 새는지 재보기
# 타깃 누수 재현 — "세그먼트별 과거 생존율" 피처 (실행하면 부풀림 크기가 나옵니다)
# ─────────────────────────────────────────────
# 동료가 만든 세그먼트 키 (운임 + 등급 조합)
from sklearn.preprocessing import TargetEncoder

seg = df["fare"].round(2).astype(str) + "_" + df["pclass"].astype(str)
print(f"세그먼트 키의 고유값: {seg.nunique()}개 / 전체 {len(seg)}명")
print(f"  → 세그먼트 하나에 평균 {len(seg) / seg.nunique():.1f}명뿐입니다. 이것이 문제의 씨앗입니다.")
print()

# ✗ 누수 버전 — 전체 데이터의 타깃으로 세그먼트별 생존율을 계산해 피처로 씀
seg_rate = y.groupby(seg).mean()            # 전체 y를 봤다!
df_leak = df.copy()
df_leak["seg_survival_rate"] = seg.map(seg_rate)

LEAK_NUM = FEAT_NUM + ["seg_survival_rate"]
pre_leak = ColumnTransformer([
    ("num", Pipeline([("impute", SimpleImputer(strategy="median")),
                      ("scale", StandardScaler())]), LEAK_NUM),
    ("cat", Pipeline([("impute", SimpleImputer(strategy="most_frequent")),
                      ("encode", OneHotEncoder(handle_unknown="ignore"))]), FEAT_CAT),
])
leaky = cross_val_score(
    Pipeline([("pre", pre_leak), ("clf", LogisticRegression(max_iter=5000, random_state=RANDOM_STATE))]),
    df_leak[LEAK_NUM + FEAT_CAT], y, cv=cv, scoring="accuracy")

# ✓ Pipeline 버전 — 같은 아이디어를 TargetEncoder로, Pipeline 안에서 (조각별로 따로 계산)
df_ok = df.copy()
df_ok["segment"] = seg
pre_ok = ColumnTransformer([
    ("num", Pipeline([("impute", SimpleImputer(strategy="median")),
                      ("scale", StandardScaler())]), FEAT_NUM),
    ("cat", Pipeline([("impute", SimpleImputer(strategy="most_frequent")),
                      ("encode", OneHotEncoder(handle_unknown="ignore"))]), FEAT_CAT),
    ("seg", TargetEncoder(random_state=RANDOM_STATE), ["segment"]),
])
clean = cross_val_score(
    Pipeline([("pre", pre_ok), ("clf", LogisticRegression(max_iter=5000, random_state=RANDOM_STATE))]),
    df_ok[FEAT_NUM + FEAT_CAT + ["segment"]], y, cv=cv, scoring="accuracy")

print(f"✗ 누수 버전 (전체 y로 생존율 계산)  CV {leaky.mean():.4f} ± {leaky.std():.4f}")
print(f"✓ Pipeline 버전 (조각별로 따로 계산) CV {clean.mean():.4f} ± {clean.std():.4f}")
print()
print(f"→ 부풀림 {leaky.mean() - clean.mean():+.4f}")
print(f"  누수 버전의 표준편차({leaky.std():.4f})가 유난히 작은 것도 누수 신호입니다.")

세그먼트 키의 고유값: 243개 / 전체 891명
  → 세그먼트 하나에 평균 3.7명뿐입니다. 이것이 문제의 씨앗입니다.

✗ 누수 버전 (전체 y로 생존율 계산)  CV 0.8945 ± 0.0074
✓ Pipeline 버전 (조각별로 따로 계산) CV 0.8350 ± 0.0165

→ 부풀림 +0.0595
  누수 버전의 표준편차(0.0074)가 유난히 작은 것도 누수 신호입니다.


c:\Users\pyostry\.conda\envs\myenv\Lib\site-packages\sklearn\preprocessing\_target_encoder.py:341: FutureWarning: `TargetEncoder.shuffle` and `TargetEncoder.random_state` are deprecated in version 1.9 and will be removed in version 1.11. Pass a cross-validation generator as `cv` argument to specify the shuffling behaviour instead.
  warnings.warn(
c:\Users\pyostry\.conda\envs\myenv\Lib\site-packages\sklearn\preprocessing\_target_encoder.py:341: FutureWarning: `TargetEncoder.shuffle` and `TargetEncoder.random_state` are deprecated in version 1.9 and will be removed in version 1.11. Pass a cross-validation generator as `cv` argument to specify the shuffling behaviour instead.
  warnings.warn(
c:\Users\pyostry\.conda\envs\myenv\Lib\site-packages\sklearn\preprocessing\_target_encoder.py:341: FutureWarning: `TargetEncoder.shuffle` and `TargetEncoder.random_state` are deprecated in version 1.9 and will be removed in version 1.11. Pass a cross-validation generator as `cv` argument to specify 

In [24]:
# ─────────────────────────────────────────────
# [C12] Part 3 · 📊 데이터로 확인해 봅시다 — 묶고, 새는지 재보기
# 전처리 누수(유형 ①)는 얼마나 부풀리나 — 정직한 재검증 (실행하면 차이가 나옵니다)
# ─────────────────────────────────────────────
# ✗ 누수 버전 — 대치와 표준화를 전체 데이터로 먼저
X_leaked = make_pre().fit_transform(X)      # 전체 X로 fit (검증 조각 포함)
leak_pre = cross_val_score(
    LogisticRegression(max_iter=5000, random_state=RANDOM_STATE),
    X_leaked, y, cv=cv, scoring="accuracy")

# ✓ Pipeline 버전
clean_pre = cross_val_score(
    Pipeline([("pre", make_pre()),
              ("clf", LogisticRegression(max_iter=5000, random_state=RANDOM_STATE))]),
    X, y, cv=cv, scoring="accuracy")

print("누수 유형별 부풀림 크기 비교")
print(f"  유형 ② 타깃 누수   : {leaky.mean() - clean.mean():+.4f}   ([C11] 결과)")
print(f"  유형 ① 전처리 누수 : {leak_pre.mean() - clean_pre.mean():+.4f}")
print()
print(f"    누수 버전   CV {leak_pre.mean():.4f} ± {leak_pre.std():.4f}")
print(f"    Pipeline    CV {clean_pre.mean():.4f} ± {clean_pre.std():.4f}")
print()
print("→ 전처리 누수는 이 데이터에서 거의 부풀리지 않았습니다. 그래도 원칙은 같습니다.")

누수 유형별 부풀림 크기 비교
  유형 ② 타깃 누수   : +0.0595   ([C11] 결과)
  유형 ① 전처리 누수 : +0.0000

    누수 버전   CV 0.8204 ± 0.0080
    Pipeline    CV 0.8204 ± 0.0080

→ 전처리 누수는 이 데이터에서 거의 부풀리지 않았습니다. 그래도 원칙은 같습니다.


In [28]:
# 1) Pipeline 만들기
pipe = Pipeline([
    ("impute", SimpleImputer(strategy="median")),
    ("scale", StandardScaler()),
    ("clf", DecisionTreeClassifier(max_depth=4, random_state=42))
])

# 2) 계층 5겹 교차검증
skf = StratifiedKFold(n_splits=5, shuffle=True, random_state=42)
scores = cross_val_score(pipe, X[FEAT_NUM], y, cv=skf, scoring="accuracy")
print(f"5겹 CV 정확도: {scores.mean():.4f} ± {scores.std():.4f}")

# 3) named_steps로 age 중앙값 꺼내기 (Pipeline을 한 번 fit해야 확인 가능)
pipe.fit(X[FEAT_NUM], y)

age_idx = FEAT_NUM.index("age")
imputer_median_age = pipe.named_steps["impute"].statistics_[age_idx]
overall_median_age = X["age"].median()

print(f"\nPipeline이 학습한 age 중앙값: {imputer_median_age:.2f}")
print(f"전체 데이터의 age 중앙값: {overall_median_age:.2f}")

5겹 CV 정확도: 0.7026 ± 0.0224

Pipeline이 학습한 age 중앙값: 28.00
전체 데이터의 age 중앙값: 28.00


1. Pipeline이 누수를 막는 메커니즘 (한 문장)
"Pipeline은 fit()이 호출될 때마다, 그 안에 포함된 모든 전처리 단계(imputer, scaler 등)를 '지금 주어진 데이터(예: 그 fold의 train)에만' 새로 학습시키고, 이후 다른 데이터(test)에는 그 학습된 기준을 transform만 적용하도록 강제한다."

2. ColumnTransformer가 필요한 이유
"숫자형 컬럼(중앙값 대치, 표준화)과 범주형 컬럼(최빈값 대치, 원-핫 인코딩)은 서로 완전히 다른 전처리 방법이 필요한데, Pipeline 하나만으로는 '컬럼마다 다른 처리'를 지정할 수 없어서, ColumnTransformer로 컬럼 그룹별로 서로 다른 파이프라인을 배정해줘야 하기 때문이다."
오늘 앞서 age(숫자)에 median 전략을 쓰다가, sex(문자)가 섞여서 에러가 났던 것도, 바로 이 "컬럼마다 다른 처리가 필요하다"는 걸 무시해서 생긴 문제였어요.

3. 타깃 누수(+0.0595)와 전처리 누수(+0.000)의 부풀림 차이가 왜 이렇게 클까요
"타깃 누수는 '정답과 사실상 같은 정보'를 모델에 직접 흘려주는 것이라 성능이 극적으로 부풀려지지만, 전처리 누수는 '평균이나 중앙값 같은 요약 통계치'만 살짜 오염시키는 것이라, 그 영향이 미미하기 때문이다."

In [30]:
# ─────────────────────────────────────────────
# [C14] Part 4 · 📊 데이터로 확인해 봅시다 — 파이프 안에서 실험하기
# Pipeline 통째로 GridSearchCV — 전처리 옵션까지 함께 (실행하면 best와 변화량이 나옵니다)
# ─────────────────────────────────────────────
from sklearn.model_selection import GridSearchCV

pipe_tune = Pipeline([("pre", make_pre()),
                      ("clf", LogisticRegression(max_iter=5000, random_state=RANDOM_STATE))])

grid = {
    "clf__C": [0.01, 0.1, 1, 10],                        # 모델 파라미터
    "pre__num__impute__strategy": ["median", "mean"],    # 전처리 옵션까지 튜닝
}

gs = GridSearchCV(pipe_tune, grid, cv=cv, scoring="accuracy", n_jobs=-1)
gs.fit(X, y)

n_comb = len(grid["clf__C"]) * len(grid["pre__num__impute__strategy"])
print(f"탐색 공간: {n_comb}조합 × 5겹 = {n_comb * 5}회 학습")
print(f"파라미터 이름 규칙: 단계명__파라미터명 (중첩은 __로 이어 붙임)")
print()
print(f"best 조합: {gs.best_params_}")
print(f"best CV : {gs.best_score_:.4f}")

base = cross_val_score(pipe_tune, X, y, cv=cv, scoring="accuracy")
print()
print(f"기본값 CV {base.mean():.4f} → 튜닝 후 {gs.best_score_:.4f}   변화 {gs.best_score_ - base.mean():+.4f}")

탐색 공간: 8조합 × 5겹 = 40회 학습
파라미터 이름 규칙: 단계명__파라미터명 (중첩은 __로 이어 붙임)

best 조합: {'clf__C': 10, 'pre__num__impute__strategy': 'mean'}
best CV : 0.8227

기본값 CV 0.8204 → 튜닝 후 0.8227   변화 +0.0022


In [31]:
# ─────────────────────────────────────────────
# [C15] Part 4 · 📊 데이터로 확인해 봅시다 — 파이프 안에서 실험하기
# 레버별 효과 비교 — 피처 vs 모델 vs 튜닝 (실행하면 각 레버의 순증이 나옵니다)
# ─────────────────────────────────────────────
def cv_of(num_cols, cat_cols, clf):
    """주어진 피처 구성과 모델의 계층 5겹 CV 정확도를 돌려줍니다."""
    pre = ColumnTransformer([
        ("num", Pipeline([("impute", SimpleImputer(strategy="median")),
                          ("scale", StandardScaler())]), num_cols),
        ("cat", Pipeline([("impute", SimpleImputer(strategy="most_frequent")),
                          ("encode", OneHotEncoder(handle_unknown="ignore"))]), cat_cols),
    ])
    return cross_val_score(Pipeline([("pre", pre), ("clf", clf)]),
                           df[num_cols + cat_cols], y, cv=cv, scoring="accuracy").mean()


LR = lambda: LogisticRegression(max_iter=5000, random_state=RANDOM_STATE)
RAW = ["pclass", "age", "sibsp", "parch", "fare"]
DERIVED = RAW + ["family_size", "is_alone", "fare_per_person"]

s1 = cv_of(RAW, ["embarked"], LR())                          # 최소 구성
s2 = cv_of(RAW, ["embarked", "sex"], LR())                   # 강한 원본 피처 추가
s3 = cv_of(DERIVED, ["embarked", "sex"], LR())               # 파생 3개 추가
s4 = cv_of(DERIVED, ["embarked", "sex", "who"], LR())        # 도메인 피처 추가
s5 = cv_of(DERIVED, ["embarked", "sex", "who"],
           RandomForestClassifier(n_estimators=300, random_state=RANDOM_STATE))  # 모델 교체

gs_rf = GridSearchCV(
    Pipeline([("pre", make_pre()), ("clf", RandomForestClassifier(random_state=RANDOM_STATE))]),
    {"clf__n_estimators": [200, 400], "clf__min_samples_leaf": [1, 3, 5],
     "clf__max_depth": [None, 8]},
    cv=cv, scoring="accuracy", n_jobs=-1).fit(X, y)
s6 = gs_rf.best_score_                                        # 튜닝

steps = [("① 수치 5개 + embarked", s1, None),
         ("② + sex", s2, s2 - s1),
         ("③ + 파생 3개", s3, s3 - s2),
         ("④ + who", s4, s4 - s3),
         ("⑤ 모델 교체 → RandomForest", s5, s5 - s4),
         ("⑥ 튜닝 (RF 12조합)", s6, s6 - s5)]
for label, val, delta in steps:
    d = f"   ({delta:+.4f})" if delta is not None else ""
    print(f"  {label:30} {val:.4f}{d}")

print()
print(f"  피처 레버 ①→④   {s4 - s1:+.4f}   ← 데이터·피처")
print(f"  모델 레버 ④→⑤   {s5 - s4:+.4f}")
print(f"  튜닝 레버 ⑤→⑥   {s6 - s5:+.4f}")
print(f"\n  RF best: {gs_rf.best_params_}")

  ① 수치 5개 + embarked             0.7026
  ② + sex                        0.7935   (+0.0909)
  ③ + 파생 3개                      0.8014   (+0.0079)
  ④ + who                        0.8204   (+0.0191)
  ⑤ 모델 교체 → RandomForest         0.8170   (-0.0034)
  ⑥ 튜닝 (RF 12조합)                 0.8395   (+0.0225)

  피처 레버 ①→④   +0.1178   ← 데이터·피처
  모델 레버 ④→⑤   -0.0034
  튜닝 레버 ⑤→⑥   +0.0225

  RF best: {'clf__max_depth': 8, 'clf__min_samples_leaf': 5, 'clf__n_estimators': 200}


In [32]:
# ─────────────────────────────────────────────
# [C16] Part 4 · 📊 데이터로 확인해 봅시다 — 파이프 안에서 실험하기
# 실험 기록표 — 조건 → 지표 → 결정 (실행하면 오늘의 실험 이력이 표로 나옵니다)
# ─────────────────────────────────────────────
experiments = pd.DataFrame([
    {"조건": "수치 5 + embarked (최소)", "CV 정확도": s1,
     "결정": "기준선으로 채택"},
    {"조건": "+ sex", "CV 정확도": s2,
     "결정": "★채택 — 단일 피처 최대 효과 (+0.091)"},
    {"조건": "+ 파생 3개(family_size·is_alone·fare_per_person)", "CV 정확도": s3,
     "결정": "채택 — 효과는 작지만(+0.008) 해석에 도움"},
    {"조건": "+ who", "CV 정확도": s4,
     "결정": "★채택 (+0.019)"},
    {"조건": "모델 교체 → RandomForest", "CV 정확도": s5,
     "결정": "기각 — 오히려 하락(-0.003). 단, 튜닝 여지 확인 위해 ⑥ 진행"},
    {"조건": "RandomForest 튜닝 12조합", "CV 정확도": s6,
     "결정": "★최종 채택 (+0.023)"},
    {"조건": "로지스틱 C·대치전략 8조합 튜닝", "CV 정확도": gs.best_score_,
     "결정": "기각 — 이득 +0.002로 표준편차 미만"},
])
print("실험 기록표 — 조건 → 지표 → 결정")
print()
print(experiments.round(4).to_string(index=False))
print()
print("→ '결정' 열이 핵심입니다. 기각한 실험도 지우지 않고 남깁니다.")
print("  다음 사람이 같은 실패를 반복하지 않게 하는 것이 기록의 목적입니다.")

실험 기록표 — 조건 → 지표 → 결정

                                           조건  CV 정확도                                       결정
                         수치 5 + embarked (최소)  0.7026                                 기준선으로 채택
                                        + sex  0.7935               ★채택 — 단일 피처 최대 효과 (+0.091)
+ 파생 3개(family_size·is_alone·fare_per_person)  0.8014              채택 — 효과는 작지만(+0.008) 해석에 도움
                                        + who  0.8204                             ★채택 (+0.019)
                         모델 교체 → RandomForest  0.8170 기각 — 오히려 하락(-0.003). 단, 튜닝 여지 확인 위해 ⑥ 진행
                         RandomForest 튜닝 12조합  0.8395                          ★최종 채택 (+0.023)
                           로지스틱 C·대치전략 8조합 튜닝  0.8227                  기각 — 이득 +0.002로 표준편차 미만

→ '결정' 열이 핵심입니다. 기각한 실험도 지우지 않고 남깁니다.
  다음 사람이 같은 실패를 반복하지 않게 하는 것이 기록의 목적입니다.


In [33]:
# ─────────────────────────────────────────────
# [C17] Part 4 · 📊 데이터로 확인해 봅시다 — 파이프 안에서 실험하기
# joblib 저장 → 재로드 → 예측 일치 확인 (실행하면 일치 여부와 파일 크기가 나옵니다)
# ─────────────────────────────────────────────
import os

best_model = gs_rf.best_estimator_
MODEL_PATH = "titanic_pipeline.joblib"

joblib.dump(best_model, MODEL_PATH)
loaded = joblib.load(MODEL_PATH)

pred_before = best_model.predict(X)
pred_after = loaded.predict(X)
same = np.array_equal(pred_before, pred_after)

print(f"저장 파일: {MODEL_PATH} ({os.path.getsize(MODEL_PATH) / 1024 / 1024:.2f} MB)")
print(f"재로드 후 예측이 원본과 동일: {same}")
print()
print("저장된 Pipeline이 품고 있는 것")
inner_pre = loaded.named_steps["pre"]
print(f"  1. 전처리 — 결측 대치 기준값, 표준화 기준, One-Hot 범주 목록")
print(f"     변환 후 피처 {len(inner_pre.get_feature_names_out())}개")
print(f"  2. 모델   — RandomForest {loaded.named_steps['clf'].n_estimators}그루")
print()
print("→ 전처리 기준까지 함께 저장되므로 새 데이터에 '학습 때와 똑같은 변환'이 적용됩니다.")
print("  전처리 코드를 따로 관리하면 이 일관성이 언젠가 반드시 깨집니다.")

저장 파일: titanic_pipeline.joblib (1.38 MB)
재로드 후 예측이 원본과 동일: True

저장된 Pipeline이 품고 있는 것
  1. 전처리 — 결측 대치 기준값, 표준화 기준, One-Hot 범주 목록
     변환 후 피처 16개
  2. 모델   — RandomForest 200그루

→ 전처리 기준까지 함께 저장되므로 새 데이터에 '학습 때와 똑같은 변환'이 적용됩니다.
  전처리 코드를 따로 관리하면 이 일관성이 언젠가 반드시 깨집니다.


In [34]:
# [C18] Part 4 · ⌨️ 백문이 불여일타! (4)
# ⌨️ 백문이 불여일타! (4) — 파생변수를 하나씩 넣어 개별 효과 재기

derived_cols = ["family_size", "is_alone", "fare_per_person"]

# 여기에 코드를 작성하세요

#[문제]
#실험을 하나 더 설계해 기록표에 추가합니다.

#1) [C15]의 cv_of 함수를 써서, 파생변수를 하나씩만 넣었을 때의 효과를 각각 측정합니다.
#  (family_size만 / is_alone만 / fare_per_person만)
#   기준은 ② 구성(RAW + embarked + sex)입니다.
#2) 세 파생변수 중 어느 것이 단독으로 가장 효과가 컸습니까?
#3) 세 효과를 각각 더한 값과 [C15]의 ③ 순증(+0.0079)을 비교합니다.
#   같습니까? 다르다면 왜 그럴까요?
s2 = cv_of(RAW, ["embarked", "sex"], LR())

# 파생변수를 하나씩만 추가했을 때의 효과 측정
derived_cols = ["family_size", "is_alone", "fare_per_person"]
individual_effects = {}

for col in derived_cols:
    score = cv_of(RAW + [col], ["embarked", "sex"], LR())
    effect = score - s2
    individual_effects[col] = effect
    print(f"{col:20} 단독 추가: {score:.4f} (효과: {effect:+.4f})")

# 2) 단독 효과가 가장 큰 파생변수
best = max(individual_effects, key=individual_effects.get)
print(f"\n단독 효과가 가장 큰 파생변수: {best} ({individual_effects[best]:+.4f})")

# 3) 세 개별 효과의 합 vs [C15]의 ③ 순증 비교
sum_of_individual = sum(individual_effects.values())
s3 = cv_of(DERIVED, ["embarked", "sex"], LR())   # [C15]의 ③과 동일한 계산
actual_net_gain = s3 - s2

print(f"\n세 개별 효과의 합:     {sum_of_individual:+.4f}")
print(f"[C15] ③ 실제 순증:     {actual_net_gain:+.4f}")
print(f"차이:                   {sum_of_individual - actual_net_gain:+.4f}")

family_size          단독 추가: 0.7935 (효과: +0.0000)
is_alone             단독 추가: 0.8014 (효과: +0.0079)
fare_per_person      단독 추가: 0.7935 (효과: -0.0000)

단독 효과가 가장 큰 파생변수: is_alone (+0.0079)

세 개별 효과의 합:     +0.0079
[C15] ③ 실제 순증:     +0.0079
차이:                   -0.0000


In [36]:
# [C19] ❓ 오늘의 퀴즈 · 코드 퀴즈
# 코드 퀴즈 — 빈칸(____)을 채워 Pipeline 통째 튜닝

quiz_pipe = Pipeline([("pre", make_pre()),
                      ("clf", DecisionTreeClassifier(random_state=RANDOM_STATE))])
quiz_grid = {
    "clf__max_depth": [3, 4, 5, 6],                       # clf 단계의 max_depth
    "pre__num__impute__strategy": ["median", "mean"], # 대치 전략
}
quiz_gs = GridSearchCV(quiz_pipe, quiz_grid, cv=cv, scoring="accuracy").fit(X, y)
print("best:", quiz_gs.best_params_, "→ CV", round(quiz_gs.best_score_, 4))

# 여기에 코드를 작성하세요

best: {'clf__max_depth': 3, 'pre__num__impute__strategy': 'median'} → CV 0.8283
